src/python/pants/backend/codegen/tasks/simple_codegen_task.py

```diff
@@ -82,22 +123,82 @@ def codegen_workdir(self, target):
     Generally, subclasses should not need to override this method. If they do, it is crucial that
     the implementation is /deterministic/ -- that is, the return value of this method should always
     be the same for the same input target.
+    :param Target target: the codegen target (e.g., a java_protobuf_library).
     :return: The absolute file path.
     """
-    # TODO(gm): This method will power the isolated/global strategies for what directories to put
-    # generated code in, once that exists. This will work in a similar fashion to the jvm_compile
-    # tasks' isolated vs global strategies, generated code per-target in a way that avoids
-    # collisions.
-    return self.workdir
+    return os.path.join(self.workdir, self._codegen_workdir_suffix(target, self.codegen_strategy))
+
+  def _execute_strategy_isolated(self, targets):
+    for target in targets:
+      self.execute_codegen([target])
+
+  def _execute_strategy_global(self, targets):
+    self.execute_codegen(targets)
+
+  def _find_sources_generated_by_target(self, target):
+    if target.id in self._generated_sources_cache:
+      for source in self._generated_sources_cache[target.id]:
+        yield source
+    target_workdir = self.codegen_workdir(target)
+    if not os.path.exists(target_workdir):
+      return
+    # breadth-first search to find generated files.
+    frontier = [target_workdir]
```

Comment:
Why did you choose to roll your own tree walk?  This looks like a job for `os.walk()`, in which case we have `safe_walk()` in dirutils.py to handle internationalization correctly which you probably want to use.